### PR TITLE
Add tests for DdotUnitary and DdotInc

### PR DESCRIPTION
**Pull request by vladimir-ch**

### CODE REVIEW

File: asm/ddot_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+			}
+		}
+	}
+}
+
+func newGuardedVector(data []float64, inc int) (v, frontGuard, backGuard []float64) {
+	if inc < 0 {
+		inc = -inc
+	}
+	guard := 2 * inc
```
Comment (kortschak) [inline asm/ddot_test.go:136]: Why is this 2*inc?

File: asm/ddot_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	} {
+		x, _, _ := newGuardedVector(test.xData, 1)
+		y, _, _ := newGuardedVector(test.yData, 1)
+		got := DdotUnitary(x, y)
+
+		if math.IsNaN(got) {
+			t.Errorf("test %v: invalid memory access in DdotUnitary", i)
+			continue
+		}
+		if got != test.want {
```
Comment (kortschak) [inline asm/ddot_test.go:53]: Check that the dots (here and below) don't write anywhere in the data slice?

File: asm/ddot_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+				if got != want {
+					t.Errorf("test %v: DdotInc failed. incX = %v, incY = %v, want %v, got %v",
+						i, incX, incY, want, got)
+				}
+			}
+		}
+	}
+}
+
+func newGuardedVector(data []float64, inc int) (v, frontGuard, backGuard []float64) {
```
Comment (kortschak) [inline asm/ddot_test.go:132]: Are you planning on using the guards for the scales?

Comment (btracey) [inline asm/ddot_test.go:132]: Add  a comment to this function please.

File: asm/ddot_test.go (hunk 1 above)
Comment (vladimir-ch) [inline asm/ddot_test.go:136]: The guard size should depend on `inc` and `1*inc` seems like too short. Should I make it even longer?

File: asm/ddot_test.go (hunk 2 above)
Comment (vladimir-ch) [inline asm/ddot_test.go:53]: I thought about it but then neglected it. You are right that the low-level code here warrants extra precaution, so I will add it.

File: asm/ddot_test.go (hunk 1 above)
Comment (kortschak) [inline asm/ddot_test.go:136]: 2*inc seems fine (you see if you are off by one, or worse - further resolution doesn't seem necessary). I was just curious if this was your reasoning.

File: asm/ddot_test.go (hunk 2 above)
Comment (vladimir-ch) [inline asm/ddot_test.go:53]: Done

File: asm/ddot_test.go (hunk 3 above)
Comment (vladimir-ch) [inline asm/ddot_test.go:132]: Done both

**Comment by vladimir-ch** [comment]
PTAL

File: asm/ddot_test.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+			yData: []float64{0, 1, 0, 3, -4, 5, -6},
+			want:  4,
+		},
+	} {
+		x, xFront, xBack := newGuardedVector(test.xData, 1)
+		y, yFront, yBack := newGuardedVector(test.yData, 1)
+		got := DdotUnitary(x, y)
+
+		if !allNaN(xFront) || !allNaN(xBack) {
+			t.Errorf("test %v: out-of-bounds write to first argument", i)
```
Comment (kortschak) [inline asm/ddot_test.go]: When you fail here, actually print the guards so it can be seen without modifying the tests - also cases below.

**Comment by kortschak** [comment]
LGTM after extra output added.

File: asm/ddot_test.go (hunk 4 above)
Comment (vladimir-ch) [inline asm/ddot_test.go]: Done

**Pull request merged**